storage/types.go

```diff
@@ -411,6 +413,40 @@ type databaseIndex interface {
 		query index.Query,
 		opts index.QueryOptions,
 	) (index.QueryResults, error)
+
+	// Close will release the index resources and close the index.
+	Close() error
+}
+
+// databaseIndexInsertQueue is a queue used in-front of the indexing component
+// for Writes. NB: this is an interface to allow easier unit tests in databaseIndex.
+type databaseIndexInsertQueue interface {
+	// Start starts accepting writes in the queue.
+	Start() error
+
+	// Stop stops accepting writes in the queue.
+	Stop() error
+
+	// Insert inserts the provided document to the index queue which processes
+	// inserts to the index asynchronously. It executes the provided callbacks
+	// based on the result of the execution. The returned wait group can be used
+	// if the insert is required to be synchronous.
+	Insert(d doc.Document, s indexInsertLifecycleHooks) (*sync.WaitGroup, error)
+}
+
+// indexInsertLifecycleHooks provides a set of callback hooks to allow the
+// databaseIndex to do lifecycle management of any resources retained
+// during indexing.
+type indexInsertLifecycleHooks interface {
```

Comment:
Yeah we can we rename this something like `OnIndexSeries`?  This would be similar to `OnRetrieveBlock(...)` and `OnReadBlock` type, etc:
https://github.com/m3db/m3db/blob/352f2364f7202e591c69017b33cf78e6b0885f8d/storage/block/types.go#L219-L230

Reply:
Sure thing.